components/script/dom/htmlformelement.rs

```diff
@@ -417,55 +409,77 @@ impl HTMLFormElement {
-
-                self.encode_plaintext(&mut form_data)
-            }
-        };
-
         // Step 18
         let win = window_from_node(self);
         match (&*scheme, method) {
-            // https://html.spec.whatwg.org/multipage/#submit-dialog
-            (_, FormMethod::FormDialog) => return, // Unimplemented
+            (_, FormMethod::FormDialog) => {
+                // TODO: Submit dialog
+                // https://html.spec.whatwg.org/multipage/#submit-dialog
+            }
             // https://html.spec.whatwg.org/multipage/#submit-mutate-action
-            ("http", FormMethod::FormGet) | ("https", FormMethod::FormGet) => {
-                // FIXME(SimonSapin): use url.query_pairs_mut() here.
-                load_data.url.set_query(Some(&*parsed_data));
+            ("http", FormMethod::FormGet) | ("https", FormMethod::FormGet) | ("data", FormMethod::FormGet) => {
+                load_data.headers.set(ContentType::form_url_encoded());
+                self.mutate_action_url(&mut form_data, &mut load_data.url);
                 self.plan_to_navigate(load_data, &win);
             }
             // https://html.spec.whatwg.org/multipage/#submit-body
             ("http", FormMethod::FormPost) | ("https", FormMethod::FormPost) => {
                 load_data.method = Method::Post;
-                load_data.data = Some(parsed_data.into_bytes());
+                self.submit_as_entity_body(&mut load_data, &mut form_data, enctype);
                 self.plan_to_navigate(load_data, &win);
             }
             // https://html.spec.whatwg.org/multipage/#submit-get-action
-            ("file", _) | ("about", _) | ("data", FormMethod::FormGet) |
+            ("file", _) | ("about", _) | ("data", FormMethod::FormPost) |
             ("ftp", _) | ("javascript", _) => {
                 self.plan_to_navigate(load_data, &win);
             }
-            _ => return // Unimplemented (data and mailto)
+            ("mailto", FormMethod::FormPost) => {
+                // TODO: Mail as body
+                // https://html.spec.whatwg.org/multipage/#submit-mailto-body
+            }
+            ("mailto", FormMethod::FormGet) => {
+                // TODO: Mail with headers
+                // https://html.spec.whatwg.org/multipage/#submit-mailto-headers
+            }
+            _ => return,
         }
     }
 
+    // https://html.spec.whatwg.org/multipage/#submit-mutate-action
+    fn mutate_action_url(&self, form_data: &mut Vec<FormDatum>, url: &mut Url) {
+        let encoding = self.pick_encoding();
+        let charset = &*encoding.whatwg_name().unwrap();
+
+        url.query_pairs_mut().clear()
+           .encoding_override(Some(self.pick_encoding()))
+           .extend_pairs(form_data.into_iter().map(|field| (field.name.clone(), field.replace_value(charset))));
+    }
+
+    // https://html.spec.whatwg.org/multipage/#submit-body
+    fn submit_as_entity_body(&self, load_data: &mut LoadData,
```

Comment:
This function (and `mutate_action_url`) is very weird because I'd expect it to also navigate to a URL, but it does not.